new-frontiers-14/frontier-station-14
# Revalidate loadouts before equipping.

## About the PR



In all cases where we equip a RoleLoadout, we explicitly validate the loadout given.

## Why / Balance


https://discord.com/channels/1123826877245694004/1176985498359713823/1313333625173049414

A player logged in as SR, has an existing character with loadout settings from a previous schema.  Since that character was saved, the PDA was moved from starting gear to role loadout (with a new group).  The saved character is missing this group from their character's loadout.

For some reason (I do not currently understand why - very much related but not fixed here), this loadout did not get validated, and the user passes a loadout that is missing this group to the function to equip it, it'll still get caught and fixed.  This also prevents some abuse in case somebody passes bogus loadouts, at the cost of some additional processing.

## How to test

0. (Optional) Nuke your local test DB.
1. Load server, client.  Make sure you can pick SR.
2. Edit your character's SR loadout and save it - give your character a hat or something to validate that the loadout is there.
4. Open a SQLite browser (https://sqlitebrowser.org/ works fine).
5. Go to the profile_loadout_group table, remove the StationRepPDA item for your character.
6. Go to the profile_loadout table, remove the SrPDA entry.
7. Close the DB connection, be sure to commit the changes to the DB.
8. Join the round.
9. On master, you should be missing your PDA.  On this branch, you should have it.

## Media


## Requirements



## Breaking changes


**Changelog**


:cl:
- fix: Fixed a case where existing characters can spawn without gear when the role changes.
